File: algorithms/sorting/pancake-sort/go/pancake_sort.go
package pancakesort

/**
 * PancakeSort implementation.
 * Sorts the array by repeatedly flipping subarrays.
 * It returns a new sorted slice without modifying the original input.
 */
func PancakeSort(arr []int) []int {
	n := len(arr)
	if n <= 1 {
		return append([]int{}, arr...)
	}

	result := make([]int, n)
	copy(result, arr)

	for currSize := n; currSize > 1; currSize-- {
		mi := findMax(result, currSize)

		if mi != currSize-1 {
			flip(result, mi)
			flip(result, currSize-1)
		}
	}

	return result
}

func flip(arr []int, k int) {
	i := 0
	for i < k {
		arr[i], arr[k] = arr[k], arr[i]
		i++
		k--
	}
}

func findMax(arr []int, n int) int {
	mi := 0
	for i := 0; i < n; i++ {
		if arr[i] > arr[mi] {
			mi = i
		}
	}
	return mi
}

File: algorithms/sorting/pancake-sort/rust/pancake_sort.rs
/**
 * Pancake Sort implementation.
 * Sorts the array by repeatedly flipping subarrays.
 */
pub fn pancake_sort(arr: &[i32]) -> Vec<i32> {
    let mut result = arr.to_vec();
    let n = result.len();

    if n <= 1 {
        return result;
    }

    for curr_size in (2..=n).rev() {
        let mi = find_max(&result, curr_size);

        if mi != curr_size - 1 {
            flip(&mut result, mi);
            flip(&mut result, curr_size - 1);
        }
    }

    result
}

fn flip(arr: &mut [i32], k: usize) {
    let mut i = 0;
    let mut j = k;
    while i < j {
        arr.swap(i, j);
        i += 1;
        j -= 1;
    }
}

fn find_max(arr: &[i32], n: usize) -> usize {
    let mut mi = 0;
    for i in 0..n {
        if arr[i] > arr[mi] {
            mi = i;
        }
    }
    mi
}
